Dihedral group representations

lkdvos opened the pull request:
This PR adds the representations of the dihedral groups (or equivalently, charge-conjugated `ZNIrrep`).
I also slightly reorganized the files since the `irreps.jl` file was getting rather long.

Jutho commented on src/irreps/dnirrep.jl:
```diff
@@ -0,0 +1,279 @@
+- `isodd::Bool`: the representation of charge conjugation.
+
+Combined these take the values ``+0, -0, 1, ..., (N - 1) / 2`` for odd ``N``, and
+``+0, -0, 1, ..., N / 2 - 1, +(N/2), -(N/2)`` for even ``N``.
```
Not sure if this needs to be even more explicit about the relation between the signs and the `isodd` boolean, as well as about the dimensionality of the representations.

Jutho commented on src/irreps/dnirrep.jl:
```diff
@@ -0,0 +1,279 @@
+    print(io, "(", a.j, ", ", a.isodd, ")")
+    return nothing
+end
+function Base.show(io::IO, ::MIME"text/plain", a::DNIrrep{N}) where {N}
```
In which context is this show method used? This would print the sector in a format that is not valid code, right?

lkdvos replied:
This would be the one used when printing to strings, because indeed it's not valid code (that's what typically the two-argument version is reserved for).
I think the added value is rather low though, so I simply removed this.

Jutho commented on src/irreps/dnirrep.jl:
```diff
@@ -0,0 +1,279 @@
+    D = fusiontensor(a, f, d)
+    D1 = reshape(view(D, :, :, 1), size(D, 1), size(D, 2))
+
+    return @tensor conj(D1[1 5]) * conj(C[2 4 5]) * A[1 2 3] * B1[3 4]
```
Well that is a bit lazy 😄

Jutho replied:
But after thinking about it, I agree that the alternative is probably horrible.

Jutho commented on src/irreps/dnirrep.jl:
```diff
@@ -0,0 +1,279 @@
+    return ifelse((c.j == 0) & c.isodd & !(a.j == b.j == 0) & !(iseven(N) && (a.j == b.j == (N >> 1))), -R, R)
+end
+
+const _invsqrt2 = 1 / sqrt(2)
```
I would move ethat one inside `fusiontensor` and make sure it is of type `sectorscalartype(I)`.

lkdvos commented:
I think I addressed all the comments here. The main remaining missing piece is to actually figure out all of the `Fsymbol` cases instead of leveraging the CGCs, but for now I would like to just leave this as-is, until it turns out that this is needed from a performance point of view.
It's definitely easy to start adding in some of the special cases gradually, so I would argue that we might as well already merge this if we are happy with the interface etc.
